connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/webhook/InboundWebhookRestController.java

```diff
@@ -89,44 +91,79 @@ private ResponseEntity<?> processWebhook(
+      connectorResponse = verify(connectorHook, payload);
+      if (connectorResponse == null) {
+        // when verification was skipped
+        // Step 2: trigger and correlate
+        var webhookResult = connectorHook.triggerWebhook(payload);
+        var ctxData = toWebhookTriggerResultContext(webhookResult);
+        connector.context().correlate(ctxData);
+        var processVariablesContext = toWebhookResultContext(webhookResult);
+        if (webhookResult.response() != null) {
+          // Step 3a: return response crafted by webhook itself
+          connectorResponse = ResponseEntity.ok(webhookResult.response().body());
         } else {
-          connectorResponse = ResponseEntity.ok().build();
+          // Step 3b: response body expression was defined and evaluated, or 200 OK otherwise
+          connectorResponse =
+              buildBodyExpressionResponseOrOk(webhookResult, processVariablesContext);
         }
       }
     } catch (Exception e) {
       LOG.info("Webhook: {} failed with exception", connector.context().getDefinition(), e);
-      if (e instanceof FeelEngineWrapperException feelEngineWrapperException) {
-        var error =
-            new FeelExpressionErrorResponse(
-                feelEngineWrapperException.getReason(), feelEngineWrapperException.getExpression());
-        connectorResponse = ResponseEntity.unprocessableEntity().body(error);
-      } else if (e instanceof ConnectorException connectorException) {
-        if (e instanceof WebhookConnectorException webhookConnectorException) {
-          connectorResponse = handleWebhookConnectorException(webhookConnectorException);
-        } else {
-          connectorResponse =
-              ResponseEntity.unprocessableEntity()
-                  .body(
-                      new ErrorResponse(
-                          connectorException.getErrorCode(), connectorException.getMessage()));
-        }
+      connectorResponse = buildErrorResponse(e);
+    }
+    return connectorResponse;
+  }
+
+  protected ResponseEntity verify(
+      WebhookConnectorExecutable connectorHook, WebhookProcessingPayload payload) throws Exception {
+    var verificationResult = connectorHook.verify(payload);
+    if (verificationResult != null) {
+      HttpHeaders headers = new HttpHeaders();
+      Optional.of(verificationResult)
+          .map(VerifiableWebhook.WebhookHttpVerificationResult::headers)
+          .ifPresent(map -> map.forEach(headers::add));
+      return new ResponseEntity<>(
+          verificationResult.body(), headers, verificationResult.statusCode());
+    }
+    return null;
+  }
+
+  protected ResponseEntity buildBodyExpressionResponseOrOk(
+      WebhookResult webhookResult, WebhookResultContext processVariablesContext) {
+    if (webhookResult.responseBodyExpression() != null) {
+      var httpResponseData = webhookResult.responseBodyExpression().apply(processVariablesContext);
```

Comment:
Couldnt this throw an exception?

Reply:
Very much depends on `apply()` implementation. The method call is covered behind `try-catch`.

Reply:
@sbuettner do you have any specific issue in mind?

Reply:
Not really but usually we catch exceptions where they happen.